Initial port of libcurl

jschlumpp opened the pull request:
This is a port of the well-known curl library to Unikraft.

marcrittinghaus committed:
Add library files

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
Adjust compilation options

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
Update to version 7.81.0

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

marcrittinghaus committed:
Disable unsupported SRP in LibreSSL

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

jschlumpp committed:
Add configuration options for most optional features

Signed-off-by: Marco Schlumpp <marco.schlumpp@gmail.com>

skuenzer committed:
Update library dependencies, requiring full-featured libc

Unfortunately, `libbcurl` can't be compiled with `nolibc` at the moment
(`fread`, `fwrite`, and co. are missing). This commit makes `libcurl` only
available if `HAVE_LIBC` was set. Additionally, VFS and socket operation
are needed by `libcurl`.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
Include dependency to `lib/posix-event`

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

jschlumpp committed:
Add more configuration options to the library

This allows to strip a few hundred kilobytes of the resulting library.

Signed-off-by: Marco Schlumpp <marco.schlumpp@gmail.com>

jviotti commented:
One of the initial things I found testing this branch on C++20 is that the includes are only for C, so `#include <curl/curl.h>` fails. This works for me:

```diff
diff --git a/Makefile.uk b/Makefile.uk
index 2f628b5..1ce969d 100644
--- a/Makefile.uk
+++ b/Makefile.uk
@@ -57,9 +57,12 @@ LIBCURL_SRC = $(LIBCURL_ORIGIN)/$(LIBCURL_BASENAME)
 # Library includes
 ################################################################################
 LIBCURL_CINCLUDES += -I$(LIBCURL_BASE)/
+LIBCURL_CXXINCLUDES += -I$(LIBCURL_BASE)/
 LIBCURL_CINCLUDES += -I$(LIBCURL_SRC)/lib
+LIBCURL_CXXINCLUDES += -I$(LIBCURL_SRC)/lib

 CINCLUDES-$(CONFIG_LIBCURL) += -I$(LIBCURL_SRC)/include
+CXXINCLUDES-$(CONFIG_LIBCURL) += -I$(LIBCURL_SRC)/include

 ################################################################################
 # Flags
```

jschlumpp committed:
Add curl to C++ include paths

The curl library can also be used from C++ directly. So add the public
library headers to the C++ include path.

Co-Authored-by: Juan Cruz Viotti <jv@jviotti.com>
Signed-off-by: Marco Schlumpp <marco.schlumpp@gmail.com>

jschlumpp committed:
Enable verbose message by default

The detailed options were introduced to end up with a very small build
of the curl library. However, this is not very user friendly and makes
it hard to figure out what went wrong if something does not work. So
keep the option but toggle the default.

Signed-off-by: Marco Schlumpp <marco.schlumpp@gmail.com>

jschlumpp commented:
I added the `CXXINCLUDES` change (the include path for the internal headers should not need a change).

Also I changed the default value of the verbose message option to make it a bit more user friendly while leaving the option to reduce the size of the library.

jviotti commented:
Awesome. Thanks!

jschlumpp committed:
Remove SIOCGIFADDR usage on LWIP

LWIP provides a incomplete definition in a public header. This breaks
the code using it.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

StefanJum commented on Config.uk:
```diff
@@ -0,0 +1,130 @@
+	bool "Enable IMAP support"
+	default n
+
+config LIBCURL_LDAP
```
Enabling this will give a `origin/curl-7.81.0/lib/ldap.c:51:11: fatal error: ldap.h: No such file or directory` error (`musl` + `lwip`), do we have any `ldap` library available for Unikraft? If so, we should add it as a dependency here.

jschlumpp replied:
I don't think we have the ldap libary packaged currently. I can remove the option if it disturbs you.

StefanJum replied:
I would comment it out so we know about it, but I wouldn't let it in if it does not build.

jschlumpp replied:
I added a dependency on the `LIBLDAP` configuration option even if we don't have that one yet. When the library is ported it most likely has this name according to the naming scheme. So this way I think the intention is more clear than just commenting it out.

StefanJum reviewed (commented):
See the one comment.
Also, do you have a working application code to test this? I keep getting `Unsupported protocol` for anything I try.

jschlumpp commented:
To generate a small example program you could try to use the curl command line: `curl --libcurl /tmp/client.c https://example.com`. That'll generate a simple program, but you might have to do some adjustments.

jschlumpp committed:
Depend on libldap for LDAP support

While the libldap library has not been ported yet, this makes libcurl
already ready for that.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

StefanJum reviewed (approved):
All good now, thanks.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

RaduNichita reviewed (approved):
All good also from my side.

Reviwed-by: Radu Nichita <radunichita99@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
